python_modules/dagster/dagster/_core/execution/asset_backfill.py

```diff
@@ -907,11 +910,37 @@ def execute_asset_backfill_iteration_inner(
         evaluation_time=backfill_start_time,
     )
 
-    run_requests = build_run_requests(
-        asset_partitions=asset_partitions_to_request,
-        asset_graph=asset_graph,
-        run_tags={**run_tags, BACKFILL_ID_TAG: backfill_id},
+    # check if all assets have backfill policies if any of them do, otherwise, raise error
+    asset_backfill_policies = [
+        asset_graph.get_backfill_policy(asset_key)
+        for asset_key in {
+            asset_partition.asset_key for asset_partition in asset_partitions_to_request
+        }
+    ]
+    all_assets_have_backfill_policies = all(
+        backfill_policy is not None for backfill_policy in asset_backfill_policies
     )
+    if all_assets_have_backfill_policies:
```

Comment:
Moved this check to be outside of `build_run_requests` so we can make a decision here to call the default build_run_requests or the newly added build_run_requests_with_backfill_policies. 
Also added "all asset must all have backfill policy together" check here and fail the backfill with DagsterBackfillFailedError if condition not met.